table/update_schema.go

```diff
@@ -0,0 +1,532 @@
+				if delete == field.ID {
+					foundInDeletes = true
+				}
+			}
+			if !foundInDeletes {
+				panic("Cannot add column, name already exists: " + name)
+			}
+
+		}
+
+		// cannot add a required column without a default value
+		if initialDefaultValue == nil && !required && !us.allowIncompatibleChanges {
+			panic("incompatible change: cannot add required column without a default value: " + name)
+		}
+
+		nestedField := &iceberg.NestedField{
+			Name:     name,
+			ID:       new_id,
+			Required: required,
+			Type:     dataType,
+			Doc:      doc,
+		}
+
+		if initialDefaultValue != nil {
+			nestedField.InitialDefault = initialDefaultValue
+		}
+
+		us.updates[new_id] = nestedField
+		us.parentToAddedIDs[-1] = append(us.parentToAddedIDs[-1], new_id)
+		us.addedNameToID[name] = new_id
+	}
+
+	return us
+}
+
+func (su *UpdateSchema) internalUpdateColumnRequirement(name string, isRequired bool) {
+	field := su.findForUpdate(name)
+	if field == nil {
+		panic("Cannot update missing column: " + name)
+	}
+
+	if (!isRequired && !field.Required) || (isRequired && field.Required) {
+		// if the change is a noop, allow it even if allowIncompatibleChanges is false
+		return
+	}
+
+	isDefaultedAdd := su.isAdded(name) && field.InitialDefault != nil
+
+	if !isRequired && !isDefaultedAdd && !su.allowIncompatibleChanges {
+		panic("Cannot change column nullability: " + name + ": optional -> required")
+	}
+
+	for _, id := range su.deletes {
+		if id == field.ID {
+			panic("Cannot update a column that will be deleted: " + field.Name)
+		}
+	}
+
+	field.Required = isRequired
+
+	su.updates[field.ID] = field
+}
+
+func (us *UpdateSchema) RenameColumn(fromPath, newName string) *UpdateSchema {
+	field := us.findField(fromPath)
+	if field == nil {
+		panic("Cannot rename missing column: " + fromPath)
+	}
+	// block rename if the column is being deleted
+	for _, id := range us.deletes {
+		if id == field.ID {
+			panic("Cannot rename a column that will be deleted: " + fromPath)
+		}
+	}
+
+	// clone or create update entry
+	upd, ok := us.updates[field.ID]
+	if !ok {
+		c := *field // copy
+		upd = &c
+	}
+
+	upd.Name = newName
+	us.updates[field.ID] = upd
+
+	// identifier set book-keeping
+	if _, ok := us.identifierFields[fromPath]; ok {
+		delete(us.identifierFields, fromPath)
+		us.identifierFields[newName] = struct{}{}
+	}
+	return us
+}
+
+// DeleteColumn removes a column from the schema.
+func (us *UpdateSchema) DeleteColumn(name string) *UpdateSchema {
+	field := us.findField(name)
+	if field == nil {
+		panic("Cannot delete missing column: " + name)
+	}
+
+	if _, ok := us.parentToAddedIDs[field.ID]; ok {
+		panic("Cannot delete a column that has additions: " + name)
+	}
+
+	if _, ok := us.updates[field.ID]; ok {
+		panic("Cannot delete a column that has updates: " + name)
+	}
+
+	us.deletes = append(us.deletes, field.ID)
+
+	return us
+}
+
+func (us *UpdateSchema) UpdateColumnType(name string, newType iceberg.Type) *UpdateSchema {
```

Comment:
do we need to have separate update functions for each element? Could we just have a single `UpdateColumn` function which takes optional arguments to decide what to update?

Reply:
i gave it a try 
please look at the current way and let me know if thats what you were thinking ?